include/swift/Remote/RemoteAddress.h

```diff
@@ -18,85 +18,274 @@
 #ifndef SWIFT_REMOTE_REMOTEADDRESS_H
 #define SWIFT_REMOTE_REMOTEADDRESS_H
 
+#include "swift/ABI/MetadataRef.h"
+#include "swift/Basic/RelativePointer.h"
+#include "llvm/ADT/DenseMapInfo.h"
+#include "llvm/ADT/Hashing.h"
+#include <cassert>
 #include <cstdint>
-#include <string>
+#include <functional>
+#include <iostream>
 #include <llvm/ADT/StringRef.h>
-#include <cassert>
+#include <ostream>
+#include <sstream>
+#include <string>
 
 namespace swift {
 namespace remote {
 
 /// An abstract address in the remote process's address space.
 class RemoteAddress {
-  uint64_t Data;
+  uint64_t Data = 0;
+  // Buffer to be used by subclasses
+  uint64_t Buffer = 0;
+
 public:
-  explicit RemoteAddress(const void *localPtr)
-    : Data(reinterpret_cast<uintptr_t>(localPtr)) {}
+  static RemoteAddress makeAddress(uint64_t addressData) {
+    static RemoteAddress Dummy;
+    return Dummy.makeAddressImpl(addressData);
+  }
 
-  explicit RemoteAddress(uint64_t addressData) : Data(addressData) {}
+  virtual RemoteAddress makeAddressImpl(uint64_t addressData) {
+    return RemoteAddress(addressData, 0);
+  }
 
-  explicit operator bool() const {
-    return Data != 0;
+  static RemoteAddress makeAddressFromPtr(const void *localPtr) {
+    return makeAddress((uint64_t)localPtr);
   }
 
-  template <class T>
-  const T *getLocalPointer() const {
-    return reinterpret_cast<const T*>(static_cast<uintptr_t>(Data));
+  explicit RemoteAddress() {}
+
+  virtual ~RemoteAddress() = default;
+
+  template <class IntegerType>
+  RemoteAddress fromThis(IntegerType data) const {
```

Comment:
Instead of this how about

`RemoteAddress(value, otherRemoteAddress.getAddressSpace())`